pyccel/parser/semantic.py

```diff
@@ -3659,88 +3659,123 @@ def _visit_For(self, expr):
             for_expr.end_annotation = expr.end_annotation
         return for_expr
 
-
     def _visit_FunctionalFor(self, expr):
-        old_index   = expr.index
-        new_index   = self.scope.get_new_name()
-        expr.substitute(old_index, new_index)
+        """
+        Visit and transform a FunctionalFor AST node into an equivalent code block.
 
-        target  = expr.expr
-        index   = new_index
-        indices = [self.scope.get_expected_name(i) for i in expr.indices]
-        dims    = []
-        body    = expr.loops[1]
+        This method processes a `FunctionalFor` expression and transforms the loop structure
+        into a corresponding code block.
 
-        idx_subs = {}
-        #scope = self.create_new_loop_scope()
+        Parameters
+        ----------
+        expr : pyccel.ast.functionalexpr.FunctionalFor
+            The FunctionalFor AST node.
+
+        Returns
+        -------
+        pyccel.ast.basic.CodeBlock
+            A code block containing the equivalent loops and necessary variable allocations for the given `FunctionalFor` expression.
+        """
+        if expr.target_type != 'list':
+            old_index   = expr.index
+            new_index   = self.scope.get_new_name()
+            expr.substitute(old_index, new_index)
+            index = new_index
 
-        # The symbols created to represent unknown valued objects are temporary
+        target  = expr.expr
+        indices = []
+        dims = []
+        idx_subs = {}
         tmp_used_names = self.scope.all_used_symbols.copy()
+        body = expr.loops[0 if expr.target_type == 'list' else 1]
         i = 0
-        while isinstance(body, For):
 
-            stop  = None
+        # Inner function to handle PythonNativeInt variables
+        def handle_int_loop_variable(var_name):
+            indices.append(var_name)
+            var = self._create_variable(var_name, PythonNativeInt(), None, {})
+            dvar = self._infer_type(var)
+            variables.append((var, dvar))
+
+        # Inner function to handle iterable variables
+        def handle_iterable_variable(var_name, element):
+            indices.append(var_name)
+            dvar = self._infer_type(element)
+            class_type = dvar.pop('class_type')
+            if class_type.rank > 0:
+                class_type = class_type.switch_rank(class_type.rank - 1)
+                dvar['shape'] = dvar['shape'][1:]
+            if class_type.rank == 0:
+                dvar['shape'] = None
+                dvar['memory_handling'] = 'stack'
+            var = self._create_variable(var_name, class_type, None, dvar)
+            dvar['class_type'] = class_type
+            variables.append((var, dvar))
```

Comment:
Why is this specific to `FunctionalFor`? Is there nothing from the `For` handling that would let you do something similar?